### PR TITLE
feat(source): support canal (espcially TiCDC) as a CDC format

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

 - add cannal json source parser that is compatible with both TiCDC and the original canal.
 - rename some modules

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Connector (sources & sinks)

### Release note

```sql
create materialized source s (
  id integer,
  name varchar,
  is_adult bool,
  reg_time timestamp,
  balance decimal,
  win_rate double,
  PRIMARY KEY (id)
) with (
  connector = 'kafka',
  topic = 'cannal_json',
  properties.bootstrap.server = '127.0.0.1:29092'
) row format CANAL_JSON;
```

### CODE REVIEW

File: src/source/src/parser/canal/operators.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub const TICDC_UPDATE_EVENT: &str = "UPDATE";
+pub const TICDC_INSERT_EVENT: &str = "INSERT";
+pub const TICDC_DELETE_EVENT: &str = "DELETE";
+pub const TICDC_QUERY_EVENT: &str = "QUERY";
```
Comment (neverchanje) [inline src/source/src/parser/canal/operators.rs:18]: Are these event types TiDB-specific? Or they are just general Canal types?

Comment (waruto210) [reply src/source/src/parser/canal/operators.rs]: They are general Canal types.

File: src/source/src/parser/canal/mod.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        }
+
+        // {
+        //     let (op, row) = rows.next().unwrap();
+        //     assert_eq!(op, Op::Insert);
+        //     assert_eq!(row.value_at(0).to_owned_datum(), Some(ScalarImpl::Int32(2)));
+        //     assert_eq!(
+        //         row.value_at(1).to_owned_datum(),
+        //         (Some(ScalarImpl::Utf8("alex".into())))
+        //     );
```
Comment (tabVersion) [inline src/source/src/parser/canal/mod.rs:149]: please remove the comments

File: src/source/src/parser/canal/simd_json_parser.rs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+use crate::parser::canal::operators::*;
+use crate::{
+    ensure_rust_type, ensure_str, ParseFuture, SourceParser, SourceStreamChunkRowWriter, WriteGuard,
+};
+
+const AFTER: &str = "data";
+const BEFORE: &str = "old";
+const OP: &str = "type";
+const ISDDL: &str = "isDdl";
```
Comment (tabVersion) [inline src/source/src/parser/canal/simd_json_parser.rs:33]: IS_DDL

**Review by tabVersion — COMMENTED** [review]
Rest LGTM

**Review by waruto210 — APPROVED** [review]
LGTM

**Comment by hengm3467** [comment]
@tabVersion @neverchanje This is not documented yet. I guess we'll need to document it. Right?

**Comment by tabVersion** [comment]
> Yes, we need to include it in the CDC page. But I'm not sure whether we should document it as TiDB CDC or Canal CDC. @tabVersion

as TiDB, I guess. We don't fully support canal but TiCDC, a dialect of canal.

**Pull request merged**